### PR TITLE
Mega refactor

### PR DESCRIPTION
**Pull request by Mpdreamz**
This PR represents @jedmao hard work [refactoring the internals of EditorConfig VS extension](https://github.com/jedmao/editorconfig-visualstudio/tree/enforce).
It copies parts of the [codemaid](http://www.codemaid.net/) architecture (with permission) in order to create a smoother experience hooking into the needed events in the editing pipeline. 

This was never properly released. Only builds distributed on issues, and once, shortly, using the wrong project id on the gallery (causing two editorconfig extensions to be available). See for more information https://github.com/editorconfig/editorconfig/issues/173

This should fix not restoring global settings properly on for projects/solutions that do not have an `editorconfig` file. 

Other changes bundled into this PR:

* Moved project to VS 2013
* Moved to .NET 4.5 and using the VS 2013 SDK
* The above means we will drop support for **VS2010** ! 
* Fixed project id so that when we publish to the gallery we won't end up with two editorconfigs :)
* Made sure the new completions and syntax highlighting still work in the refactored project.
* Moved to [managed .net editorconfig core](https://github.com/editorconfig/editorconfig-core-net) which passes the original c core's tests and several additional [C# ones](https://github.com/editorconfig/editorconfig-core-net/tree/master/src/EditorConfig.Tests). This makes building the project that much more easier. 
* Finally added an `EditorConfig file template`, works on `Add New Item` on solution root or solution folders and will create an `.editorconfig` file next to the solution file.

image

It relies on a hack, showing an ugly filename to the user, to create a file with only an extension (which is not allowed in the dialog)
image

but will always ignore whatever is passed and always create a file called `.editorconfig`

image

I **really hope** to make time to set up a test box for VS2012 and VS2013, so far I have tested this against VS 2015 Community. This PR might linger a while since I'll be on a 3 week holiday a week from now. If you have VS2012 and VS2013 please help by trying this 

http://editorconfig.blob.core.windows.net/vsix/EditorConfig.vsix

and reporting results back on this ticket

ping @jedmao @nulltoken @swansontec @KindDragon

### CODE REVIEW

**Comment by Mpdreamz** [comment]
- This branch will try to clean readonly files. 
- Clean on open should be turned off by default IMO.

File: CONTRIBUTORS [hunk 1]
```diff
@@ -5,3 +5,5 @@ Martijn Laarman
 Hong Xu
 Tom Potts
 Ethan J. Brown
+Jed Hunsaker
```
Comment (jedmao) [inline CONTRIBUTORS]: My name has changed, legally, to Jed Mao since then.

**Comment by jedmao** [comment]
:+1: I agree that clean on open should be removed.

**Comment by Mpdreamz** [comment]
*Clean up on open file is gone
*Opened file's settings are always outputted in the EditorConfig output pane. 

The restore of global settings is too aggressive on this branch:

* [ ]  Refactorings from e.g resharper do not take editorconfig settings into account. 
* [ ] Changing indent_size while using tabs can have a very jumpy effect when the current editor receives and loses focus (e.g when jumping to solution explorer).

Going to experiment using restoring settings on closing the editor as suppose to on lost focus

**Comment by jedmao** [comment]
@Mpdreamz the reason global settings are restored when a tab loses focus is so they can edit their global settings w/o being confused about what settings they are editing.

**Comment by Mpdreamz** [comment]
thanks for posting the original reason @jedmao :+1: 

This conflicts with losing focus because of a refactor dialog popping up (either resharper or visual studio's own). Kind of deadlocked here haha :)

In my opinion having refactorings respect editor settings is more important then a confusing text editor settings dialog in visual studio (which is already pretty confusing if you ask me).

The  prior you will hit more often then the latter and we can only fix one AFAIK.

**Comment by jedmao** [comment]
As long as you still restore the EditorConfig settings on tab focus, I support your strategy.

**Comment by pruiz** [comment]
Are there any pending issues impeding merge of this PR? :?

**Comment by Mpdreamz** [comment]
yes refactoring (roslyn based or resharper based) does not take editorconfig settings in to account because we reset very aggressively. 

I've been running every day with the extension (0.6 as linked in this thread) now and VS2015 often crashes, this could be because of other extensions or vs 2015 RC itself. I'm now (as of today) running on the published version of editorconfig again to compare how often VS 2015 crashes. 

Definitely not ready to pull the trigger and publish this yet though.

**Comment by ctlajoie** [comment]
Is this still progressing? I found a bug related to file paths that I'd like to submit a PR for, but it looks like it would conflict with this one if I did...

**Comment by Mpdreamz** [comment]
@ctlajoie I want to get back to this but I have and will be too swamped to spend brain cycles on this. 

The main gripe I have with this branch that needs fixing is that it restores settings to aggressively so that when you do new file or any refactoring it does not take into account the editorconfig settings until you an explicit code-format.

**Comment by ctlajoie** [comment]
@Mpdreamz Can I humbly suggest that this PR be merged so anyone else (if so inclined) can fix the problems? I have noticed a few issues with this version myself (none severe though), but have avoided looking for the problem until this gets merged so I can potentially submit my own PR.

**Comment by jedmao** [comment]
Aren't these "issues" pre-existing anyway? Then what are we losing by merging this PR in?

**Comment by Manuzor** [comment]
So, what's the status of this? It sounds very exciting and I'm looking forward to an improved VS2015 experience with EditorConfig!

**Comment by drauch** [comment]
We are currently evaluating whether we want to use editorconfig with Visual Studio, however, we are not sure whether the project is still maintained. Is this branch being merged anytime soon or is editorconfig-visualstudio abandoned?

Sorry for the harsh words, I'm sure you have a ton of stuff to do and this is only a spare-time project for the community, still, if you don't have time, please say so, maybe someone else wants to fill in or at least we know, that it doesn't make sense to wait for this PR to be merged.

**Comment by Mpdreamz** [comment]
@drauch I am itching to jump back on it. I have no intend to give up on this branch just been swamped with deliverables at work for the past months. I have moved back to the gallert published version of editorconfig (in vs2015) because I got bit by the refactorings not taking settings into account bug on this branch and that was impeding my work. 

I am going to merge this in as is, so we can move forward on this.

**Comment by dg9ngf** [comment]
Is this a fix to #18? If so, that's definitely interesting as that bug is a blocker for me.

Do I understand it right that settings are currently restored whenever the editor loses focus? And that it was discussed to change that to when the editor is closed instead? Sounds a lot more reasonable to me. What about multiple editors open in parallel? Can they conflict? Say if two files from different projects with different editorconfig settings are open.

How does the restoring work? Would the original user preferences be lost forever when Visual Studio crashes while a file with different settings is open?

**Comment by drauch** [comment]
@dg9ngf: :+1: would like to know answers to all your questions as well!

**Comment by dg9ngf** [comment]
@Mpdreamz Any news on this? I see project activity, but this blocking bug is not progressing anyhow. What's your roadmap or definition of priorities? Features before bugs? I'd love to use this, but I can't if it messes up my IDE.

**Comment by sandcastle** [comment]
@Mpdreamz How can we help to move this forward? It has been months with no love 😿 💔

**Comment by ygoe** [comment]
@sandcastle Just wait for Visual Studio "15", it will have EditorConfig support included. No need for this abandoned broken extension anymore! :-) See https://visualstudio.uservoice.com/forums/121579-visual-studio-2015/suggestions/6146845-support-editorconfig

**Comment by sandcastle** [comment]
@ygoe that's great news, do you know if it's already in the pre-RTM version atm?

**Comment by ygoe** [comment]
@sandcastle There's a link to a video in their announcement. I've watched that and it looks like this feature is going to be in the next preview (> 4) or whatever comes next. I haven't heard of an estimated RTM release date yet but expect it to be sometime in Q1/2017.

**Comment by Mpdreamz** [comment]
@ygoe  `"this abandoned broken extension"` :-1:  wording matters this is a labor of love project in our spare time. 

Both published and refactored versions have different interoperabillity problems that are hard to tackle. The published version just has the least intrusive problems and works well for many people.

VS15 getting editorconfig support takes some of the pressure of and will be much welcomed! It will extend in to roslyn formatting settings too so good times ahead! :smile:

**Comment by ygoe** [comment]
@Mpdreamz I'm sorry that my description sounds bad to you, nonetheless I consider it valid (at its time) based on public observations: this extension has (or had? we don't know) a major bug that I personally could not accept and nobody cared to respond. Communication has a big effect on how something is perceived, and the lack of communication doesn't improve my trust.

**Pull request merged**